Allow users to configure no_connect_privileged_ports

NotTsunami opened the pull request:
Don't hardcode lt::settings_pack::no_connect_privileged_ports. We support the peer alert for it already, so connections blocked by this setting should be logged without any further changes.

NotTsunami commented:
Will probably cause a merge conflict with #12176 due to same placement, but I will fix that depending on which is merged first (I believe both are the correct setting, and the final order being: Additional connections per IP, Privileged Ports, HTTPS Trackers).

Now, the real question is, why did we hardcode this value before? We support the peer alert if a peer is blocked due to privileged ports. Was there a historic reason?

Chocobo1 commented:
>Now, the real question is, why did we hardcode this value before? 

Check the links in https://github.com/qbittorrent/qBittorrent/issues/12408#issuecomment-609426501

>Don't hardcode lt::settings_pack::no_connect_privileged_ports. 

Why would a user care about this option? If the rationale is to prevent some amplification attack, then a few users enabling this setting is basically worthless and most users won't ever bother about it. Also setting `no_connect_privileged_ports=true` by default is not sensible IMO and there was multiple discussion before.

NotTsunami commented:
> > Now, the real question is, why did we hardcode this value before?
> 
> Check the links in [#12408 (comment)](https://github.com/qbittorrent/qBittorrent/issues/12408#issuecomment-609426501)
> 
> > Don't hardcode lt::settings_pack::no_connect_privileged_ports.
> 
> Why would a user care about this option? If the rationale is to prevent some amplification attack, then a few users enabling this setting is basically worthless and most users won't ever bother about it. Also setting `no_connect_privileged_ports=true` by default is not sensible IMO and there was multiple discussion before.

The inspiration of this PR was more so the frustration of connecting to privileged peers repeatedly than resisting any DoS attacks. But reading previous discussions, I see both sides. If you feel it is useless, feel free to close this.

Chocobo1 commented:
>The inspiration of this PR was more so the frustration of connecting to privileged peers repeatedly 

I wouldn't say this is totally worthless, I see exposing this setting can ease some user's mind however I don't think it just stops here.
/rant If you linger on the issue board long enough, you'll notice somebody will eventually step out and push that we change the default to the opposite and come up with endless debates, even when the topic has been discussed numerous times in the past and you see #12408 is just another instance of history repeat itself. And the worst part is they are willing to debate to death and say they *feel* the old default is obsolete and the situation is different now. I'll be honest I'm sick of it, I even thought of asking upstream to remove the knob entirely...
That said, I'm neutral of having this PR and if we really have it, I'll be very reluctant to change its default value unless there are something truly wrong about it.

FranciscoPombal commented:
@Chocobo1 @NotTsunami 
Right now I would actually be against this patch, due to the fact that it is of questionable usefulness now after recent changes in libtorrent. I elaborate on this here: https://github.com/qbittorrent/qBittorrent/issues/12408#issuecomment-617679441

Copy-pasted for convenience:

>Since the patch seems to work as intended, after reading through this once again, I have changed my mind about whether blocking connections to privileged ports should be the default.
>
>In fact, I don't think this should be a configurable settings at all unless someone can prove the following is a big deal affecting many users (I don't like the idea of making changes in qBittorrent just to appease anti-virus software, unless it is really needed):
>
>>I think one pragmatic point is that some anti-malware/anti-virus software really doesn’t like processes making connections to destination port 25 (SMTP)

Chocobo1 commented on src/gui/advancedsettings.cpp:
```diff
@@ -498,6 +501,9 @@ void AdvancedSettings::loadAdvancedSettings()
     addRow(MULTI_CONNECTIONS_PER_IP, tr("Allow multiple connections from the same IP address"), &m_checkBoxMultiConnectionsPerIp);
+    // Disable connection to privileged ports (< 1024)
+    m_checkBoxPrivilegedPortsDisabled.setChecked(session->isConnectToPrivilegedPortsDisabled());
+    addRow(DISABLE_PRIVILEGED_PORTS, (tr("Disable connecting to privileged ports") + ' ' + makeLink("https://libtorrent.org/single-page-ref.html#no_connect_privileged_ports", "(?)")), &m_checkBoxPrivilegedPortsDisabled);
```
maybe "Disable connection to peers on privileged port"?

NotTsunami replied:
I went with `Disable connection to peers on privileged ports`

thalieht commented:
Personally i don't like negative options (not sure if there is a term for it) i.e. i would prefer something like "Allow connecting to peers on privileged ports" and checked by default, rather than "Disable connection to peers on privileged ports" unchecked by default.

NotTsunami commented:
> Personally i don't like negative options (not sure if there is a term for it) i.e. i would prefer something like "Allow connecting to peers on privileged ports" and checked by default, rather than "Disable connection to peers on privileged ports" unchecked by default.

How do you feel about the patch in general? I think this should wait on being merged until more things are libtorrent side, as they may render this patch unnecessary.

Also, I think there's plenty of different options on wording, which do you all think is the best?
- `Allow connecting to peers on privileged ports` (Positive)
- `Block peer connections on privileged ports` (Negative)
- `Filter connections to privileged ports` (Negative)
- `Enable peer connections on privileged ports` (Positive)

FranciscoPombal commented:
@NotTsunami 

If this is implemented, I would prefer `Positive`, `ON` by default:
`Allow connecting to privileged ports on peers` (slightly different than any option you provided, but I think it makes it clearer that it's the destination port we're talking about).

> How do you feel about the patch in general? I think this should wait on being merged until more things are libtorrent side, as they may render this patch unnecessary.

I think it is important to clarify the scope of this patch: it does not fix the handshake bugs that lead to "fake peers" persisting connections or frequently re-establishing them. That is something only libtorrent can fix, and that seems to be going well.

It would only be fair to say that this patch fixes the "fake peers" issue if a proper libtorrent-side fix were deemed impossible. Then, this patch would be a "fix" for the "fake peers" issue insofar as it would be a client-side workaround for a library limitation.
As you said:
> The inspiration of this PR was more so the frustration of connecting to privileged peers

If the underlying cause is fixed, the workaround isn't needed.

So, under the assumption that the libtorrent fixes are enough to fully address the "fake peers" issue, which I think is a fair assumption, this patch is simply an implementation of a feature that some users requested for their use case, and should be judged as such.

With that said, here are the pros/cons that have some objective technical justification that I can see for this patch:

- Pros:
    - some antiviruses might complain about qBittorrent making connections to privileged ports (though this seems to be a complain from where back when, unclear if it happens now at all...); this patch would provide and escape hatch for that.
    - the ability to turn this off provides peace of mind for those who want to turn it off because "oh noes, privileged ports!".
- Cons:
    - some user might turn off the option (accidentally or due to "oh noes, privileged ports!") and then complain about less connections/poor performance.

Let me know if you think of more. Other than the anti-virus escape hatch which is dubious at best, I can't think of good reasons to implement this. After all, as far as the BitTorrent protocol goes, it does not care about the "privilege" of the port it connects to or receives connections from. But then again, if others push for it, I won't stand in the way.

thalieht commented:
>How do you feel about the patch in general?

I don't know enough about the matter to oppose it and since there are people who want to block these ports no matter their reason, i don't see why not.

>which do you all think is the best?

The 1st or the 2nd which doesn't "negative". Maybe it's about "do this" rather than "don't do this" or perhaps just the word "disable"... argh i don't know what's my problem.

LordNyriox commented:
@NotTsunami:  Personally, my own vote is in favor of the second option you listed: 

> `Block peer connections on privileged ports` 

"Block" just seems like a better descriptor to me than "disable" or "allow", in this case.

NotTsunami commented:
I've let this sit here for a while to see how libtorrent progresses. Do we still see this as a beneficial option? If so, I'll rebase and update the phrasing.

FranciscoPombal commented:
@NotTsunami My opinion is still https://github.com/qbittorrent/qBittorrent/pull/12577#issuecomment-619305312, which this PR seems to align with so far. So it's all good as far as I'm concerned.

Seeker2 commented:
A question I have about how this works is...
Whether this also blocks incoming connections on privileged ports or only blocks outgoing connections to privileged ports?

uTP peers and seeds tend to reuse their listening port as their outgoing port as well, but TCP almost always does not.
So an incoming TCP connection could report a privileged port as its listening port for minor confusion as to whether this feature is actually working.

NotTsunami commented:
Rebased and switched to thaliet's suggestion of `Allow connecting to peers on privileged ports`.

> A question I have about how this works is...
> Whether this also blocks incoming connections on privileged ports or only blocks outgoing connections to privileged ports?
> 
> uTP peers and seeds tend to reuse their listening port as their outgoing port as well, but TCP almost always does not.
> So an incoming TCP connection could report a privileged port as its listening port for minor confusion as to whether this feature is actually working.

According to Arvid's [comment](https://github.com/qbittorrent/qBittorrent/issues/12408#issuecomment-618853244) in the relevant issue, this would block outgoing connections to servers listening on privileged ports.

NotTsunami commented:
This last patchset is inverted, I will upload a patchset tomorrow fixing that and likely returning to the negative wording to match the libtorrent option and have the most straightforward meaning.

NotTsunami commented:
I went to submit another PR and realized I hadn't pushed from my remote branch. Rebased and opted for LordNyriox's suggestion as it matches the case of `no_connect_privileged_ports`.

Chocobo1 commented on src/webui/api/appcontroller.cpp:
```diff
@@ -749,6 +751,9 @@ void AppController::setPreferencesAction()
     if (hasKey("validate_https_tracker_certificate"))
         session->setValidateHTTPSTrackerCertificate(it.value().toBool());
+    // Block peer connections on privileged ports
+    if (hasKey("no_connect_privileged_ports"))
```
Using the same key `block_peers_on_privileged_ports` should make it less surprising for webapi users.

NotTsunami replied:
Sure, done.

Chocobo1 commented on src/webui/www/private/views/preferences.html:
```diff
@@ -2266,6 +2275,7 @@
             settings.set('utp_tcp_mixed_mode', $('utpTCPMixedModeAlgorithm').getProperty('value'));
             settings.set('enable_multi_connections_from_same_ip', $('allowMultipleConnectionsFromTheSameIPAddress').getProperty('checked'));
             settings.set('validate_https_tracker_certificate', $('validateHTTPSTrackerCertificate').getProperty('checked'));
+            settings.set('no_connect_privileged_ports', $('blockPeersOnPrivilegedPorts').getProperty('checked'));
```
don't forget this one.

Chocobo1 commented on src/gui/advancedsettings.cpp:
```diff
@@ -537,6 +540,9 @@ void AdvancedSettings::loadAdvancedSettings()
 #endif
+    // Block peer connections on privileged ports
+    m_checkBoxBlockPeersOnPrivilegedPorts.setChecked(session->blockPeersOnPrivilegedPorts());
+    addRow(BLOCK_PEERS_ON_PRIVILEGED_PORTS, (tr("Block peer connections on privileged ports") + ' ' + makeLink("https://libtorrent.org/single-page-ref.html#no_connect_privileged_ports", "(?)")), &m_checkBoxBlockPeersOnPrivilegedPorts);
```
I would suggest: "Disallow connections to peers on privileged port".

NotTsunami replied:
I prefer Nyriox's suggestion, but others are free to chime in.

Chocobo1 replied:
I'll explain myself a bit more.
Personally I think "block" means some thing is initiated but being cut off in the middle, while "disallow" means it is not permitted in the first place.

Second, IMO "Block peer connections on privileged ports" would sound similar to "don't connect to peers from our local privileged ports", of course that is nonsense but I would still like to clear that ambiguity.

Seeker2 replied:
"Block peer connections on privileged ports" also sounds like incoming peer connections using their privileged ports trying to connect to your listening port that's not a privileged port.
TCP connections are likely to use ephemeral ports, but uTP may very well use the same port as the listening port...so this issue is a real.

The elephant-in-the-room is the purpose of this feature in the first place -- to reduce DDoS-Attack potential of qBitTorrent, particularly if fed "target" ip + port addresses by malicious peers or trackers.

NotTsunami replied:
Do you suggest a different name for the functions/variables as well?

glassez replied:
>IMO "Block peer connections on privileged ports" would sound similar to "don't connect to peers from our local privileged ports", of course that is nonsense but I would still like to clear that ambiguity.

Block peers that use privileged port.
Block peer connections from privileged ports.
>Personally I think "block" means some thing is initiated but being cut off in the middle, while "disallow" means it is not permitted in the first place.

"Blocked" is currently used when peer is disallowed by IP filter.

Chocobo1 replied:
>Do you suggest a different name for the functions/variables as well?

No, they are fine.

NotTsunami replied:
Alright, adapted to Chocobo's suggestion.

LordNyriox commented on src/gui/advancedsettings.cpp:
```diff
@@ -537,6 +540,9 @@ void AdvancedSettings::loadAdvancedSettings()
 #endif
+    // Disallow connection to peers on privileged port
+    m_checkBoxBlockPeersOnPrivilegedPorts.setChecked(session->blockPeersOnPrivilegedPorts());
+    addRow(BLOCK_PEERS_ON_PRIVILEGED_PORTS, (tr("Disallow connection to peers on privileged port") + ' ' + makeLink("https://libtorrent.org/single-page-ref.html#no_connect_privileged_ports", "(?)")), &m_checkBoxBlockPeersOnPrivilegedPorts);
```
@NotTsunami:  English grammar correction: 

> Disallow connection to peers on privileged port 

…should be… 

> Disallow connection to peers on privileged **ports** 

The double plural is necessary because the two nouns are on different levels of the statement.

Chocobo1 commented:
@NotTsunami 
Thank you!
